internal/config/ociartifact/ociartifact.go

```diff
@@ -24,54 +28,135 @@ func New() Impl {
+	ref, err := docker.NewReference(name)
 	if err != nil {
-		return nil, fmt.Errorf("get container storage: %w", err)
+		return nil, fmt.Errorf("create docker reference: %w", err)
 	}
 
-	runtime, err := libimage.RuntimeFromStore(store, &libimage.RuntimeOptions{SystemContext: sys})
+	src, err := ref.NewImageSource(ctx, opts.SystemContext)
 	if err != nil {
-		return nil, fmt.Errorf("create libimage runtime: %w", err)
+		return nil, fmt.Errorf("build image source: %w", err)
 	}
 
-	images, err := runtime.Pull(ctx, ref, config.PullPolicyAlways, &libimage.PullOptions{})
+	manifestBytes, mimeType, err := src.GetManifest(ctx, nil)
 	if err != nil {
-		return nil, fmt.Errorf("pull OCI artifact: %w", err)
+		return nil, fmt.Errorf("get manifest: %w", err)
+	}
+
+	parsedManifest, err := manifest.FromBlob(manifestBytes, mimeType)
+	if err != nil {
+		return nil, fmt.Errorf("parse manifest: %w", err)
+	}
+	if opts.EnforceConfigMediaType != "" && parsedManifest.ConfigInfo().MediaType != opts.EnforceConfigMediaType {
+		return nil, fmt.Errorf(
+			"wrong config media type %q, requires %q",
+			parsedManifest.ConfigInfo().MediaType, opts.EnforceConfigMediaType,
+		)
+	}
+
+	layers := parsedManifest.LayerInfos()
+	if len(layers) < 1 {
+		return nil, errors.New("artifact needs at least one layer")
+	}
+
+	// Just supporting one layer here. This can be later enhanced by extending
+	// the PullOptions.
+	layer := layers[0]
+
+	bic := blobinfocache.DefaultCache(opts.SystemContext)
+	rc, _, err := src.GetBlob(ctx, layer.BlobInfo, bic)
+	if err != nil {
+		return nil, fmt.Errorf("get layer blob: %w", err)
+	}
+	defer rc.Close()
+
+	maxArtifactSize := defaultMaxArtifactSize
+	if opts.MaxSize > 0 {
+		maxArtifactSize = opts.MaxSize
 	}
-	image := images[0]
 
-	mountPath, err := image.Mount(ctx, nil, "")
+	layerBytes, err := readLimit(rc, maxArtifactSize)
 	if err != nil {
-		return nil, fmt.Errorf("mount OCI artifact: %w", err)
+		return nil, fmt.Errorf("read with limit: %w", err)
 	}
 
-	cleanup := func() {
-		if err := image.Unmount(true); err != nil {
-			log.Warnf(ctx, "Unable to unmount OCI artifact path %s: %v", mountPath, err)
-		}
+	if err := verifyDigest(&layer, layerBytes); err != nil {
+		return nil, fmt.Errorf("verify digest of layer: %w", err)
 	}
 
 	return &Artifact{
-		MountPath: mountPath,
-		Cleanup:   cleanup,
+		Data: layerBytes,
 	}, nil
 }
+
+func readLimit(reader io.Reader, limit int) ([]byte, error) {
+	limitedReader := io.LimitReader(reader, int64(limit+1))
+
+	res, err := io.ReadAll(limitedReader)
+	if err != nil {
+		return nil, fmt.Errorf("read from limit reader: %w", err)
+	}
+
+	if len(res) > limit {
+		return nil, fmt.Errorf("exceeded maximum allowed size of %d bytes", limit)
+	}
+
+	return res, nil
+}
+
+func verifyDigest(layer *manifest.LayerInfo, layerBytes []byte) error {
+	expectedDigest := layer.BlobInfo.Digest
+	if err := expectedDigest.Validate(); err != nil {
+		return fmt.Errorf("invalid digest %s: %w", expectedDigest, err)
```

Comment:
```suggestion
		return fmt.Errorf("invalid digest %q: %w", expectedDigest, err)
```
it may be empty or contain control characters

Reply:
Changed as suggested.